src/nnet3/nnet-discriminative-example.h

```diff
@@ -0,0 +1,410 @@
+                    // frames).  We may still need to include those frames in
+                    // the computation in order to get correct posteriors for
+                    // other parts of the lattice.
+
+  bool split; // if false, we won't split at all.
+
+  bool excise; // if false, we will skip the "excise" step.
+  
+  SplitDiscriminativeExampleConfig():
+      max_length(1024), criterion("smbr"), collapse_transition_ids(true),
+      determinize(true), minimize(true), test(false), drop_frames(false),
+      split(true), excise(true) { }
+
+  void Register(OptionsItf *opts) {
+
+    opts->Register("max-length", &max_length, "Maximum length allowed for any "
+                   "segment (i.e. max #frames for any example");
+    opts->Register("criterion", &criterion, "Criterion, 'mmi'|'mpfe'|'smbr'. "
+                   "Determines which frames may be dropped from lattices.");
+    opts->Register("collapse-transition-ids", &collapse_transition_ids,
+                   "This option included for debugging purposes");
+    opts->Register("determinize", &determinize, "If true, we determinize "
+                   "lattices (as Lattice) before splitting and possibly minimize");
+    opts->Register("minimize", &minimize, "If true, we push and "
+                   "minimize lattices (as Lattice) before splitting");
+    opts->Register("test", &test, "If true, activate self-testing code.");
+    // See "Sequence-discriminative training of deep neural networks", Vesely et al,
+    // ICASSP 2013 for explanation of frame dropping.
+    opts->Register("drop-frames", &drop_frames, "For MMI, if true we drop frames "
+                   "with no overlap of num and den frames");
+    opts->Register("split", &split, "Set to false to disable lattice-splitting.");
+    opts->Register("excise", &excise, "Set to false to disable excising un-needed "
+                   "frames (option included for debug purposes)");
+  }
+};
+
+/// This struct exists only for diagnostic purposes.  Note: the stats assume
+/// that you call SplitDiscriminative and ExciseDiscriminativeExample in the
+/// same program, and the info printed out will be wrong if this is not the
+/// case... this isn't ideal but it was more convenient.
+struct SplitExampleStats {
+  int32 num_lattices;
+  int32 longest_lattice;
+  int32 num_segments;
+  int32 num_kept_segments;
+  int64 num_frames_orig;
+  int64 num_frames_must_keep;
+  int64 num_frames_kept_after_split;
+  int32 longest_segment_after_split;
+  int64 num_frames_kept_after_excise;
+  int32 longest_segment_after_excise;
+  
+  SplitExampleStats() { memset(this, 0, sizeof(*this)); }
+  void Print();
+};
+
+/** Split a "discriminative example" into multiple pieces,
+    splitting where the lattice has "pinch points".
+    Uses "name" as the supervision object that is used to 
+    determine the "pinch points".
+ */
+void SplitDiscriminativeExample(
+    const std::string &name,
```

Comment:
I have some time to work on this project.

If we are not looking at pinch points and creating only fixed-length segments, then it should not be too difficult to have to support multiple supervision objects. This might be useful in some situations like training with MMI and CE objective. Microsoft had done this in one of their papers to fix some issues like getting large number of deletions and insertions, which we usually have.

If we use lattice forward and backward scores, would we need to update these scores during some of the training iterations since they change when the model gets updated? 

Vimal